src/transport/http.ts

```diff
@@ -0,0 +1,278 @@
+import { RequestOptions } from '../request';
+import { SubscriptionListeners } from '../subscription';
+import { XhrReadyState, ElementsHeaders, responseToHeadersObject, ErrorResponse, NetworkError } from '../network';
+import { SubscriptionEvent, SubscriptionTransport } from '../subscription';
+import { Logger } from '../logger';
+
+export enum HttpTransportState {
+  UNOPENED = 0, // haven't called xhr.send()
+  OPENING,      // called xhr.send(); not yet received response headers
+  OPEN,         // received response headers; called onOpen(); expecting message
+  ENDING,       // received EOS message; response not yet finished
+  ENDED         // called onEnd() or onError(err)
+};
+
+export default class HttpTransport implements SubscriptionTransport {
+  private xhr: XMLHttpRequest;
+  private state: HttpTransportState = HttpTransportState.UNOPENED;
+  private listeners: SubscriptionListeners;
+  private baseURL: string;
+
+  constructor(host: string, encrypted?: boolean) {
+    this.baseURL = `${encrypted !== false ? "https" : "http"}://${host}`;
+  }
+
+  public request (requestOptions: RequestOptions): XMLHttpRequest {
+    return this.createXHR(this.baseURL, requestOptions);
+  }
+
+  public subscribe(
+    path: string,
+    listeners: SubscriptionListeners,
+    headers: ElementsHeaders
+  ): number {
+    this.listeners = listeners;
+    
+    const requestOptions: RequestOptions = {
+      method: "SUBSCRIBE",
+      path: path,
+      headers: headers
+    };
+
+    this.xhr = this.createXHR(this.baseURL, requestOptions);
+
+    this.xhr.onreadystatechange = () => {
+      switch (this.xhr.readyState) {
+        case XhrReadyState.UNSENT:
+        case XhrReadyState.OPENED:
+        case XhrReadyState.HEADERS_RECEIVED:
+          this.assertStateIsIn(HttpTransportState.OPENING);
+          break;
+
+        case XhrReadyState.LOADING:
+          this.onLoading();
+          break;
+
+        case XhrReadyState.DONE:
+          this.onDone();
+          break;
+      }
+    };
+    this.state = HttpTransportState.OPENING;
+    this.xhr.send();
+
+    return 1;
+  }
+
+  public unsubscribe(subscribtionId: number): void {
```

Comment:
Because subscription ids leaked from the websocket transport to higher levels of abstractions, you have to pass them to HTTP subscriptions, but HTTP does not have subscription ids. This method should not take any arguments.